From 969ae29c3c1a12f6311631e2fe1d6d50af5240eb Mon Sep 17 00:00:00 2001
From: Mao Minkai <maominkai@wxiat.com>
Date: Tue, 7 Jun 2022 17:41:33 +0800
Subject: [PATCH] sw64: vdso: fix time calculation
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56P0Z

--------------------------------

Switch type of wall to monotonic time to u64 so we can get the
correct time.

Signed-off-by: Mao Minkai <maominkai@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/include/asm/vdso.h          | 4 ++--
 arch/sw_64/kernel/vdso/vgettimeofday.c | 8 ++++----
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/arch/sw_64/include/asm/vdso.h b/arch/sw_64/include/asm/vdso.h
index 8ecd5add42ad..7a2e23c648f3 100644
--- a/arch/sw_64/include/asm/vdso.h
+++ b/arch/sw_64/include/asm/vdso.h
@@ -41,8 +41,8 @@
 struct vdso_data {
 	u64 xtime_sec;
 	u64 xtime_nsec;
-	u32 wall_to_mono_sec;
-	u32 wall_to_mono_nsec;
+	u64 wall_to_mono_sec;
+	u64 wall_to_mono_nsec;
 	u32 cs_shift;
 	u32 cs_mult;
 	u64 cs_cycle_last;
diff --git a/arch/sw_64/kernel/vdso/vgettimeofday.c b/arch/sw_64/kernel/vdso/vgettimeofday.c
index e45618f9f2de..3579e0eddfe2 100644
--- a/arch/sw_64/kernel/vdso/vgettimeofday.c
+++ b/arch/sw_64/kernel/vdso/vgettimeofday.c
@@ -55,8 +55,8 @@ static __always_inline int do_monotonic_coarse(struct timespec64 *ts,
 		const struct vdso_data *data)
 {
 	u32 start_seq;
-	u32 to_mono_sec;
-	u32 to_mono_nsec;
+	u64 to_mono_sec;
+	u64 to_mono_nsec;
 
 	do {
 		start_seq = vdso_data_read_begin(data);
@@ -124,8 +124,8 @@ static __always_inline int do_monotonic(struct timespec64 *ts,
 {
 	u32 start_seq;
 	u64 ns;
-	u32 to_mono_sec;
-	u32 to_mono_nsec;
+	u64 to_mono_sec;
+	u64 to_mono_nsec;
 
 	do {
 		start_seq = vdso_data_read_begin(data);
-- 
2.34.1

